### ISSUE TITLE
Pod health probes are not passing through the proxy

### ISSUE THREAD

**Issue by kyessenov** [istio/manager#474]
If a pod declares a pod health probe using HTTP GET method, then the proxy drops the health check. The problem is that the pod health probe can use any port on the pod, not necessarily defined by the associated service or by the pod container ports. The ideal solution would be to simply not trap health probes, but it is not clear how to identify health check packets. 

The probes are defined per pod, meaning the services model does not have to be extended with the information about each service instance health ports - a service instance will never need to know about another instance's health port.

Therefore, I suggest we simply add a flag to the proxy agent to open specific ports and populate this flag in kube-inject script. For example, if the pod defines livenessProbe on port 3333, we will need to pass "3333" to the proxy, and the proxy can open this port if it is not already opened.

**Comment by kyessenov** [comment]
Thoughts? @ayj @andraxylia @rshriram

**Comment by kyessenov** [comment]
It seems we cannot identify the health probes easily. These are sent by node agent (kubelet) here: https://github.com/kubernetes/kubernetes/blob/master/pkg/probe/http/http.go#L61

**Comment by ayj** [comment]
Looks like we would need to pass  `path` as well as `port and do so for both liveliness and readiness probes, right? 

I'm assuming "open this port" means updating envoy configuration to pass through HTTP health checks? If so, how do we reconcile `path/port` combinations with existing routes from RDS that are coming from remote discovery service?

**Comment by kyessenov** [comment]
We don't need path in the probe. In fact, there are TCP liveness checks as well. If there is already a service port used for health checking, then it should work as-is since the HTTP route adds POD_IP:POD_PORT as one of the possible hosts. Am I missing something?

**Comment by ayj** [comment]
I was thinking health checks needed to be HTTP (not HTTPS), but maybe I'm confusing that with L4 health checks for GCLB. But assuming HTTPS, how does that work with mTLS and Istio CA? Also, do we need to exclude health checks from things like quota/access checks?

**Comment by kyessenov** [comment]
Good point. I'll file an issue regarding auth + health probes problem.

Without auth, if the port is not used by the service, then we will use tcp_proxy on that pod_ip, port combination. If the port is already used, then we do not do anything, and trust that the port is handled correctly (e.g. if the port is declared to be HTTP port, then the liveness probe will not attempt to use non-HTTP probes). 

The app has to handle the path in the probe correctly, we just need to let the packets through the proxy.

If the port is used for quota, then it should not matter who initiates the connection. That means the kubelet probes will count towards the quota, and that needs to be documented (e.g. do not use service ports for health checks guideline). @mandarjog

**Comment by kyessenov** [comment]
#479

**Comment by rshriram** [comment]
do they neec to pass thru proxy? if pod has 3 containers, can't I define multiple probes (one per component), such that a pod gets registered only when all components (app, proxy, etc) come up ?

**Comment by markvincze** [comment]
I see that the respective PR was merged, but it's not part of the `0.4.0` release.
Is there a way to get the liveness probe working already today?
Because it is not working for me with the `istio-proxy` container in place, and if I try to `curl` the liveness endpoint from one of the k8s nodes, I get a `404` from envoy:

```
HTTP/1.1 404 Not Found
date: Thu, 04 Jan 2018 14:44:43 GMT
server: envoy
content-length: 0
```

(But if I do the same curl with `localhost` from inside the container, then it works.)

This is probably the same issue, right? Is there today a known workaround for this?
